### PR TITLE
Imagenet

### PR DESCRIPTION
**Pull request by acocac**
- Working prototype with ImageNet pretrained models available from [Classification models Zoo - Keras (and TensorFlow Keras) repo](https://github.com/qubvel/classification_models)
- Requires a better handling of model arguments, `model_type`, `input_shape`, `weights`, etc in `ImageNet class` and `YAML` file.

### CODE REVIEW

File: scivision_test_plugin/model.py [hunk 1]
```diff
@@ -1,6 +1,13 @@
+import os
+
+os.environ['TF_KERAS'] = '1'
+
 import numpy as np
 from skimage.filters import gaussian
 
+if os.environ.get('TF_KERAS'):
+    from classification_models.tfkeras import Classifiers
```
Comment (quantumjot) [inline scivision_test_plugin/model.py]: Is this environmental variable necessary?

File: scivision_test_plugin/model.py [hunk 2]
```diff
@@ -15,13 +22,29 @@ def __init__(self):
...
 
     def do_something(
-        self,
-        preserve_range: bool = True,
-        sigma: float = 10.0,
-        image: np.ndarray = np.empty((64, 64, 1)),
+            self,
+            preserve_range: bool = True,
+            sigma: float = 10.0,
+            image: np.ndarray = np.empty((64, 64, 1)),
```
Comment (quantumjot) [inline scivision_test_plugin/model.py]:
```suggestion
        self,
        preserve_range: bool = True,
        sigma: float = 10.0,
        image: np.ndarray = np.empty((64, 64, 1)),
```

File: scivision_test_plugin/model.py [hunk 3]
```diff
@@ -15,13 +22,29 @@ def __init__(self):
...
     ) -> np.ndarray:
         return gaussian(image, sigma=sigma, preserve_range=preserve_range)
 
 
+class ImageNetModel:
+    def __init__(self):
+        self.pretrainedModel, self.preprocess_input = Classifiers.get('resnet18')
```
Comment (quantumjot) [inline scivision_test_plugin/model.py]:
```suggestion
    def __init__(self, model_name:str = "resnet18"):
        model, self.preprocess_input = Classifiers.get(model_name)
        
        # can build the model here, so we can reuse the prediction function
        self.pretrained_model = model(
            input_shape=(224, 224, 3), weights="imagenet", classes=1000
        )
```

File: scivision_test_plugin/model.py [hunk 4]
```diff
@@ -15,13 +22,29 @@ def __init__(self):
...
+
+    def predict(self, image: np.ndarray) -> np.ndarray:
+        model = self.pretrainedModel(
+            input_shape=(224, 224, 3), weights="imagenet", classes=1000
+        )
```
Comment (quantumjot) [inline scivision_test_plugin/model.py]: We shouldn't need to build the model each time?

```suggestion
```

File: .scivision-config_imagenet.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+name: test
+url: https://github.com/acocac/scivision-test-plugin.git
+import: scivision_test_plugin
+model: ImageNetModel
+args: None
```
Comment (quantumjot) [inline .scivision-config_imagenet.yaml]: We could add the model name in the config file:

```suggestion
args:
    model_name: resnet18
```

**Review by quantumjot — CHANGES_REQUESTED** [review]
Hi @acocac - looks good! I've made a few suggestions here. I think we will still need to change the syntax of the config file, but this will get us started.

File: scivision_test_plugin/model.py (hunk 3 above)
Comment (quantumjot) [reply scivision_test_plugin/model.py]: Also, `input_shape`, `weights` and `classes` could become arguments here. We could specify those in the config file in the same way as `model_name`

File: scivision_test_plugin/model.py (hunk 1 above)
Comment (acocac) [reply scivision_test_plugin/model.py]: You're right the environmental variable might be redundant. We can remove it.

File: scivision_test_plugin/model.py (hunk 3 above)
Comment (acocac) [reply scivision_test_plugin/model.py]: great, it looks ok for me.

File: scivision_test_plugin/model.py (hunk 4 above)
Comment (acocac) [reply scivision_test_plugin/model.py]: for the prediction function, do we have to replace `model.predict()` by `self.pretrainedModel.predict()`, am I right?

File: .scivision-config_imagenet.yaml (hunk 5 above)
Comment (acocac) [reply .scivision-config_imagenet.yaml]: Cool! It's now clear for me how to declare model arguments in the YAML file.

**Comment by acocac** [comment]
@quantumjot, thanks for the feedback, I implemented the changes suggested, please check them.

**Review by quantumjot — APPROVED** [review]

**Comment by quantumjot** [comment]
Great work @acocac - I think there will be a few issues, but we can fix those in a follow-up PR.

**Pull request merged**